src/app/libp2p_helper/src/codanet.go

```diff
@@ -418,36 +424,101 @@ func (gs *CodaGatingState) MarkPrivateAddrAsKnown(addr ma.Multiaddr) {
 	}
 }
 
-func (gs *CodaGatingState) isPeerTrusted(p peer.ID) bool {
-	return gs.TrustedPeers.Contains(p)
+type connectionAllowance int
+
+const (
+	Undecided connectionAllowance = iota
+	DenyUnknownPrivateAddress
+	DenyBannedPeer
+	DenyBannedAddress
+	Accept
+)
+
+var connectionAllowanceStrings = map[connectionAllowance]string{
+	Undecided:                 "Undecided",
+	DenyUnknownPrivateAddress: "DenyUnknownPrivateAddress",
+	DenyBannedPeer:            "DenyBannedPeer",
+	DenyBannedAddress:         "DenyBannedAddress",
+	Accept:                    "Allow",
+}
+
+func (ca connectionAllowance) String() string {
+	return connectionAllowanceStrings[ca]
+}
+
+func (c connectionAllowance) isDeny() bool {
+	return !(c == Accept || c == Undecided)
 }
 
-func (gs *CodaGatingState) isPeerBanned(p peer.ID) bool {
-	return gs.BannedPeers.Contains(p)
+func (gs *CodaGatingState) checkPeerTrusted(p peer.ID) connectionAllowance {
+	if gs.TrustedPeers.Contains(p) {
+		return Accept
+	}
+	return Undecided
+}
+
+func (gs *CodaGatingState) checkPeerBanned(p peer.ID) connectionAllowance {
+	if gs.BannedPeers.Contains(p) {
+		return DenyBannedPeer
+	}
+	return Undecided
+}
+
+// bothAccept makes sure neither allowance denies the connection
+func bothAccept(a connectionAllowance, b connectionAllowance) connectionAllowance {
+	if a == Undecided {
+		return b
+	}
+	if a == Accept {
+		if b == Undecided {
+			return Accept
+		}
+		return b
+	}
+	return a
+}
+
+// eitherAccepts makes sure either allowance allows the connection (or both are undecided)
+func eitherAccepts(a connectionAllowance, b connectionAllowance) connectionAllowance {
+	if a == Undecided {
```

Comment:
This seems incorrect given the comment. Shouldn't this be
```go
if a == Accept {
  return a
}
return b
```

Currently the behaviour seems to be
* if `a` rejects, reject
* if `a` is undecided and `b` rejects, reject
* otherwise, use `a`'s acceptance.

Alternatively, this could be renamed
```go
// unlessUndecided(a, b) returns `a` unless it is undecided, in which case it falls back to `b`
func unlessUndecided(a connectionAllowance, b connectionAllowance) connectionAllowance {
  ...
}
```

Reply:
@mrmr1993 nice catch. You can check that the logic is fine: new code implements exactly the same policy as the old code.

Naming is misleading, agree. I will follow your proposal